File: solutions/go/blackjack/1/blackjack.go
package blackjack

// ParseCard returns the integer value of a card following blackjack ruleset.
func ParseCard(card string) int {
	switch card {
        case "ace":
          return 11;
        case "jack", "queen", "king", "ten":
          return 10;
        case "nine":
          return 9;
        case "eight":
          return 8;
        case "seven":
          return 7;
        case "six":
          return 6;
        case "five":
          return 5;
        case "four":
          return 4;
        case "three":
          return 3;
        case "two":
          return 2;
        default:
          return 0;
    }
}

// FirstTurn returns the decision for the first turn, given two cards of the
// player and one card of the dealer.
func FirstTurn(card1, card2, dealerCard string) string {
	var sum = ParseCard(card1) + ParseCard(card2);
    var dc = ParseCard(dealerCard);
    
    switch {
        case sum == 22:
    		return "P";
        case sum == 21 && dc < 10: 
    		return "W";
        case sum == 21 && dc >= 10:
    		return "S";
        case sum >= 17 && sum <= 20:
    		return "S";
        case sum >= 12 && sum <= 16 && dc < 7:
    		return "S";
        default:
        	return "H";
    }
}

File: solutions/r/blackjack/1/blackjack.R
library(dplyr)

parse_card <- function(card) {
  recode_values(
    card,
    "ace" ~ 11,
    c("ten", "jack", "queen", "king") ~ 10,
    "nine" ~ 9,
    "eight" ~ 8,
    "seven" ~ 7,
    "six" ~ 6,
    "five" ~ 5,
    "four" ~ 4,
    "three" ~ 3,
    "two" ~ 2,
    default = 0
  )
}

first_turn <- function(card1, card2, dealer_card) {
  sum_cards <- parse_card(card1) + parse_card(card2)
  case_when(
    card1 == "ace" && card2 == "ace" ~ "P",
    sum_cards == 21 && parse_card(dealer_card) < 10 ~ "W",
    sum_cards == 21 && parse_card(dealer_card) >= 10 ~ "S",
    between(sum_cards, 17, 20) ~ "S",
    between(sum_cards, 12, 16) && parse_card(dealer_card) < 7 ~ "S",
    between(sum_cards, 12, 16) && parse_card(dealer_card) >= 7 ~ "H",
    sum_cards <= 11 ~ "H"
  )
}
